keps/sig-apps/2185-random-pod-select-on-replicaset-downscale/README.md

```diff
@@ -0,0 +1,442 @@
+there are a number of reasons why we don't need to preserve such behavior as is:
+
+- The behavior is not documented in the API or user docs, only in code.
+- The heuristic is applied last, after other criteria have been applied. In
+  particular, the heuristic doesn't apply when multiple Pods are in the same
+  Node or if a particular Pod has had several container restarts. While users
+  can enforce one Pod per Node with certain scheduling features, there is no
+  workaround for other criteria.
+- We are not proposing to entirely remove the heuristic, just make it more
+  laxed.
+- We are introducing a [related feature](git.k8s.io/enhancements/keps/sig-apps/1828-delete-priority-annotations)
+  that provides higher guarantees for downscaling order that users can migrate
+  to.
+  
+  <!--
+  If not merged yet, the KEP can be found in https://github.com/kubernetes/enhancements/pull/1828
+  -->
+
+## Design Details
+
+We propose a randomized approach to the algorithm for Pod victim selection
+during ReplicaSet downscale:
+
+1. Do a random shuffle of ReplicaSet Pods.
+2. Obtain wall time, and add it to [`ActivePodsWithRanks`](https://github.com/kubernetes/kubernetes/blob/dc39ab2417bfddcec37be4011131c59921fdbe98/pkg/controller/controller_utils.go#L815)
+2. Call sorting algorithm with a modified time comparison for start and
+   creation timestamp.
+
+Instead of directly comparing timestamps, the algorithm compares the elapsed
+times since the timestamp until the current time but in a logarithmic scale,
+floor rounded. This has the effect of treating elapsed times as equals when they
+have the same scale. That is, Pods that have been running for a few nanoseconds
+are equal, but they are different from pods that have been running for a few
+seconds or a few days.
+
+For example, let's assume the base 10 is used, then we have the following
+mapping for different durations:
+
+| Duration | Scale |
+|----------|-------|
+| 5ns      | 0     |
+| 23ns     | 1     |
+| 71ns     | 1     |
+| 1ms      | 6     |
+| 8ms      | 6     |
+| 50ms     | 7     |
+| 2m       | 11    |
+| 11m      | 11    |
+
+An alternative interpretation for the base 10 is that, if a Pod has been running
+for more than 10 times the time as another Pod, then the second Pod would be
+deleted first.
+
+While base 10 is quite intuitive, it might be too aggressive on bucketing
+timestamps together. A base of 2 could be similarly intuitive and provide a
+better bucketing. But if documentation is not a problem, the natural base is
+a good choice as well.
+
+### Test Plan
+
+Unit and e2e tests will be helpful to ensure continuing performance of the
+intended behavior. However, due to the random nature of downscale selection
+within each rounded bucket it will be important to keep in mind the difficulty
+in expecting a specific pod to be deleted when multiple could still be valid candidates.
+Understanding this while writing tests will significantly reduce flakes.
```

Comment:
Should we specify a high level scenario that we want to test? perhaps similar to the story.

Reply:
Added a reference to the user story with a brief description (don't think it's necessary to explicitly re-describe the entire scenario)